contracts/crowdfund/CollectionBatchBuyCrowdfund.sol

```diff
@@ -142,65 +171,80 @@ contract CollectionBatchBuyCrowdfund is BuyCrowdfundBase {
+                if (price > maxPrice) {
+                    revert MaximumPriceError(price, maxPrice);
+                }
+
+                // Add the price to the total value used for the call.
+                callValue += price;
             }
 
-            // Used to ensure no ETH is spent if the call fails.
             uint256 balanceBefore = address(this).balance;
+            {
+                // Execute the call to buy the NFT.
+                (bool success, ) = _buy(token, call.target, callValue, call.data);
 
-            // Execute the call to buy the NFT.
-            (bool success, bytes memory revertData) = _buy(
-                token,
-                args.tokenIds[i],
-                args.callTargets[i],
-                args.callValues[i],
-                args.callDatas[i]
-            );
+                if (!success) continue;
+            }
 
-            if (!success) {
-                if (args.minTokensBought >= args.tokenIds.length) {
-                    // If the call failed with revert data, revert with that data.
-                    if (revertData.length > 0) {
-                        revertData.rawRevert();
-                    } else {
-                        revert FailedToBuyNFTError(token, args.tokenIds[i]);
-                    }
-                } else {
-                    // If the call failed, ensure no ETH was spent and skip this NFT.
-                    if (address(this).balance != balanceBefore) {
-                        revert ContributionsSpentForFailedBuyError();
+            {
+                uint96 ethUsed;
+                for (uint256 j; j < call.tokensToBuy.length; ++j) {
+                    uint256 tokenId = call.tokensToBuy[j].tokenId;
+                    uint96 price = call.tokensToBuy[j].price;
+
+                    // Check whether the NFT was successfully bought.
+                    if (token.safeOwnerOf(tokenId) == address(this)) {
+                        ethUsed += price;
+                        ++tokensBought;
+
+                        // Add the token to the list of tokens to finalize.
+                        tokens[tokensBought - 1] = token;
+                        tokenIds[tokensBought - 1] = tokenId;
                     }
+                }
 
-                    continue;
+                // Check ETH spent for call is what was expected.
+                uint256 actualEthUsed = balanceBefore - address(this).balance;
+                if (ethUsed != actualEthUsed) {
+                    revert EthUsedForFailedBuyError(ethUsed, actualEthUsed);
                 }
-            }
 
-            totalEthUsed += args.callValues[i];
+                totalEthUsed += ethUsed;
+            }
 
-            ++tokensBought;
-            tokens[tokensBought - 1] = token;
-            args.tokenIds[tokensBought - 1] = args.tokenIds[i];
+            // Check if number of tokens bought has reached maximum.
+            if (tokensBought == args.maxTokensBought) break;
```

Comment:
The strict equality here might be a problem. Could this potentially exceed `maxTokensBought`, since tokens are processed in discrete batches? For example, if `maxTokensBought` is 3 and we buy two batches of `[1, 2]`, `[3, 4]`, we could still buy 4 tokens and exceed `maxTokensBought`. 

That assumes 1) the host intentionally passed more tokens than `maxTokensBought` and 2) prices are such that we somehow bought 4 tokens for the price of 3, but seems like an edge case to consider.

Reply:
Yeah this would be an issue. I've just replaced it with `numOfTokens` since it cannot be enforced in cases where multiple tokens are bought in a single call.